solana-developers/interface-pocs
# Xray

## Summary

xray is a modified solana-test-validator paired with program code-changes to support dynamic account-resolution. It is meant to solve interfaces for programs need to CPI to other programs based on account-data. 

This POC accomplishes this in two ways:
1. Modified Solana-test-validator
  i. Informs programs that they're running in special simulation (which allows programs to perform expensive logging)
  ii. Listens to logs to determine missing account pubkeys + locations
  iii. Fetches account information from a specific cluster & replays instruction with inserted AccountInfo
2. Modified program logging 
  i. Programs check `volatile_read` of static variable which determines whether to do expensive logging

## Setup

- Depends on [ngundotra/solana fork](https://github.com/solana-labs/solana/pull/29814)
- Symlink `target/deploy/solana-test-validator` as `xray` into `account-resolution/test-validator`
( `ln -s <Path>/target/deploy/solana-test-validator xray` )

## Testing

1. `cargo-build-sbf` `hello-world` example
5. Load the hello world program into NORMAL `solana-test-validator -r --bpf-program <pubkey> <path_to_program.so>`
6. Download the account data (rn expected to be `account-resolution/client/mocks/program.info`)
  i. `solana account <program pubkey> --output json -o client/mocks/program.info`
7. Flip the program's ELF bytes & execute xray via `cargo run`-ing `account-resolution/test-validator` dir
8. Start listening to logs (`solana config set -ul && solana logs`)
9. Send test transaction by running `yarn start` in `account-resolution/client`

## Debugging FAQ

- How can I tell if the `.runtime.txt` section exists in the `hello-world.so`?

Download [readelf.py](https://github.com/eliben/pyelftools) tool & execute `readelf.py target/deploy/hello_world.so -x .runtime.txt`

Should get output like:
```
Hex dump of section '.runtime.txt':
  0x00008090 00000000 68770000 07000000 00000000 ....hw..........
```

Then you can check the ELF raw bytes with `readelf.py target/deploy/hello_world.so -p .rodata | head`

Which should show you "RUNTIME" & a hex address
```
Hex dump of section '.rodata':
  0x00004488 52554e54 494d4552 756e7469 6d652069 RUNTIMERuntime i
  ...
```
